File: python/algorithms/dp/dp.py
"""
Dynamic Programming Algorithms
===============================
Classic DP patterns for SDE interview preparation.

Each function includes docstrings with time/space complexity and a
concrete demonstration in the __main__ block at the bottom.
"""

import bisect
from functools import lru_cache


# ---------------------------------------------------------------------------
# 1. Fibonacci — memoization / tabulation / space-optimized
# ---------------------------------------------------------------------------

def fibonacci(n: int) -> dict:
    """Compute the n-th Fibonacci number using three DP strategies.

    Compares memoization (top-down), tabulation (bottom-up), and
    the space-optimized two-variable approach.

    Parameters
    ----------
    n : int
        Non-negative integer index (F(0)=0, F(1)=1).

    Returns
    -------
    dict with keys:
        'memoization'    : int  — result via top-down memoization
        'tabulation'     : int  — result via bottom-up table
        'space_optimized': int  — result via O(1) space iteration

    Complexity
    ----------
    All three variants:
        Time  : O(n)
        Space : O(n) for memoization and tabulation, O(1) for optimized
    """

    # --- top-down with memoization ---
    memo: dict[int, int] = {}

    def _memo(k: int) -> int:
        if k <= 1:
            return k
        if k not in memo:
            memo[k] = _memo(k - 1) + _memo(k - 2)
        return memo[k]

    result_memo = _memo(n)

    # --- bottom-up tabulation ---
    if n == 0:
        result_tab = 0
    elif n == 1:
        result_tab = 1
    else:
        table = [0] * (n + 1)
        table[1] = 1
        for i in range(2, n + 1):
            table[i] = table[i - 1] + table[i - 2]
        result_tab = table[n]

    # --- space-optimized ---
    a, b = 0, 1
    for _ in range(n):
        a, b = b, a + b
    result_space = a

    return {
        "memoization": result_memo,
        "tabulation": result_tab,
        "space_optimized": result_space,
    }


# ---------------------------------------------------------------------------
# 2. 0/1 Knapsack
# ---------------------------------------------------------------------------

def knapsack_01(
    weights: list[int],
    values: list[int],
    capacity: int,
) -> tuple[int, list[int]]:
    """0/1 Knapsack — maximize total value without exceeding capacity.

    Each item can be taken at most once (0/1 choice).

    Parameters
    ----------
    weights : list[int]
        Weight of each item.
    values : list[int]
        Value of each item.
    capacity : int
        Maximum weight the knapsack can hold.

    Returns
    -------
    max_value : int
        Maximum achievable value.
    chosen_items : list[int]
        Indices (0-based) of items included in the optimal solution.

    Complexity
    ----------
    Time  : O(n * capacity)
    Space : O(n * capacity)  — full table kept for backtracking
    """
    n = len(weights)
    # dp[i][w] = max value using items 0..i-1 with capacity w
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        wi, vi = weights[i - 1], values[i - 1]
        for w in range(capacity + 1):
            dp[i][w] = dp[i - 1][w]
            if wi <= w:
                dp[i][w] = max(dp[i][w], dp[i - 1][w - wi] + vi)

    # Backtrack to find chosen items
    chosen = []
    w = capacity
    for i in range(n, 0, -1):
        if dp[i][w] != dp[i - 1][w]:
            chosen.append(i - 1)
            w -= weights[i - 1]
    chosen.reverse()

    return dp[n][capacity], chosen


# ---------------------------------------------------------------------------
# 3. Longest Common Subsequence
# ---------------------------------------------------------------------------

def lcs(s1: str, s2: str) -> tuple[int, str]:
    """Longest Common Subsequence of two strings.

    Parameters
    ----------
    s1, s2 : str
        Input strings.

    Returns
    -------
    length : int
        Length of the LCS.
    subsequence : str
        One LCS (there may be multiple of equal length).

    Complexity
    ----------
    Time  : O(m * n)   where m = len(s1), n = len(s2)
    Space : O(m * n)
    """
    m, n = len(s1), len(s2)
    # dp[i][j] = LCS length of s1[:i] and s2[:j]
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    # Backtrack
    seq = []
    i, j = m, n
    while i > 0 and j > 0:
        if s1[i - 1] == s2[j - 1]:
            seq.append(s1[i - 1])
            i -= 1
            j -= 1
        elif dp[i - 1][j] >= dp[i][j - 1]:
            i -= 1
        else:
            j -= 1
    seq.reverse()

    return dp[m][n], "".join(seq)


# ---------------------------------------------------------------------------
# 4. Longest Increasing Subsequence — O(n log n) patience sorting
# ---------------------------------------------------------------------------

def lis(arr: list[int]) -> tuple[int, list[int]]:
    """Longest Increasing Subsequence via patience sorting.

    Uses binary search (bisect_left) to maintain a *tails* array where
    tails[i] is the smallest tail element of all increasing subsequences
    of length i+1 found so far.

    Parameters
    ----------
    arr : list[int]
        Input sequence.

    Returns
    -------
    length : int
        Length of the LIS.
    subsequence : list[int]
        One actual LIS (reconstructed via predecessor links).

    Complexity
    ----------
    Time  : O(n log n)
    Space : O(n)
    """
    if not arr:
        return 0, []

    n = len(arr)
    tails: list[int] = []   # tails[i] = smallest tail for LIS of length i+1
    tail_idx: list[int] = []  # index in arr of each tail
    pred = [-1] * n          # predecessor array for reconstruction
    pos = [-1] * n           # position in tails array for each element

    for i, x in enumerate(arr):
        lo = bisect.bisect_left(tails, x)
        if lo == len(tails):
            tails.append(x)
            tail_idx.append(i)
        else:
            tails[lo] = x
            tail_idx[lo] = i
        pos[i] = lo
        pred[i] = tail_idx[lo - 1] if lo > 0 else -1

    # Reconstruct
    length = len(tails)
    idx = tail_idx[length - 1]
    subseq = []
    while idx != -1:
        subseq.append(arr[idx])
        idx = pred[idx]
    subseq.reverse()

    return length, subseq


# ---------------------------------------------------------------------------
# 5. Edit Distance (Levenshtein)
# ---------------------------------------------------------------------------

def edit_distance(s1: str, s2: str) -> tuple[int, list[str]]:
    """Levenshtein edit distance between two strings.

    Operations: insert, delete, substitute (each costs 1).

    Parameters
    ----------
    s1, s2 : str
        Source and target strings.

    Returns
    -------
    distance : int
        Minimum number of single-character edits.
    operations : list[str]
        Human-readable sequence of edit operations to transform s1 -> s2.

    Complexity
    ----------
    Time  : O(m * n)
    Space : O(m * n)
    """
    m, n = len(s1), len(s2)
    # dp[i][j] = edit distance between s1[:i] and s2[:j]
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(m + 1):
        dp[i][0] = i
    for j in range(n + 1):
        dp[0][j] = j

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = 1 + min(
                    dp[i - 1][j],      # delete from s1
                    dp[i][j - 1],      # insert into s1
                    dp[i - 1][j - 1],  # substitute
                )

    # Backtrack to recover operations
    ops = []
    i, j = m, n
    while i > 0 or j > 0:
        if i > 0 and j > 0 and s1[i - 1] == s2[j - 1]:
            ops.append(f"keep '{s1[i-1]}'")
            i -= 1
            j -= 1
        elif j > 0 and (i == 0 or dp[i][j - 1] <= dp[i - 1][j] and dp[i][j - 1] <= dp[i - 1][j - 1]):
            ops.append(f"insert '{s2[j-1]}'")
            j -= 1
        elif i > 0 and (j == 0 or dp[i - 1][j] <= dp[i][j - 1] and dp[i - 1][j] <= dp[i - 1][j - 1]):
            ops.append(f"delete '{s1[i-1]}'")
            i -= 1
        else:
            ops.append(f"substitute '{s1[i-1]}' -> '{s2[j-1]}'")
            i -= 1
            j -= 1
    ops.reverse()

    return dp[m][n], ops


# ---------------------------------------------------------------------------
# 6. Coin Change — minimum coins
# ---------------------------------------------------------------------------

def coin_change(coins: list[int], amount: int) -> tuple[int, list[int]]:
    """Minimum number of coins to make up *amount*.

    Parameters
    ----------
    coins : list[int]
        Available coin denominations (positive integers).
    amount : int
        Target amount (>= 0).

    Returns
    -------
    count : int
        Minimum number of coins, or -1 if the amount cannot be made.
    combination : list[int]
        Coins used in the optimal solution (multiset as sorted list).

    Complexity
    ----------
    Time  : O(amount * len(coins))
    Space : O(amount)
    """
    INF = float("inf")
    dp = [INF] * (amount + 1)
    dp[0] = 0
    last_coin = [-1] * (amount + 1)  # for reconstruction

    for a in range(1, amount + 1):
        for c in coins:
            if c <= a and dp[a - c] + 1 < dp[a]:
                dp[a] = dp[a - c] + 1
                last_coin[a] = c

    if dp[amount] == INF:
        return -1, []

    # Reconstruct
    combination = []
    cur = amount
    while cur > 0:
        combination.append(last_coin[cur])
        cur -= last_coin[cur]

    return dp[amount], sorted(combination)


# ---------------------------------------------------------------------------
# 7. Matrix Chain Multiplication
# ---------------------------------------------------------------------------

def matrix_chain_mult(dims: list[int]) -> tuple[int, str]:
    """Optimal parenthesization for matrix chain multiplication.

    Parameters
    ----------
    dims : list[int]
        Sequence of n+1 dimensions: matrix i has size dims[i] x dims[i+1].
        Must have at least 2 elements (1 matrix).

    Returns
    -------
    min_ops : int
        Minimum number of scalar multiplications.
    parenthesization : str
        Optimal parenthesization as a string, e.g. '((A0 x A1) x A2)'.

    Complexity
    ----------
    Time  : O(n^3)
    Space : O(n^2)
    """
    n = len(dims) - 1  # number of matrices
    # dp[i][j] = min cost to multiply matrices i..j (0-indexed)
    dp = [[0] * n for _ in range(n)]
    split = [[0] * n for _ in range(n)]

    for length in range(2, n + 1):         # chain length
        for i in range(n - length + 1):
            j = i + length - 1
            dp[i][j] = float("inf")
            for k in range(i, j):
                cost = dp[i][k] + dp[k + 1][j] + dims[i] * dims[k + 1] * dims[j + 1]
                if cost < dp[i][j]:
                    dp[i][j] = cost
                    split[i][j] = k

    def _build(i, j) -> str:
        if i == j:
            return f"A{i}"
        k = split[i][j]
        return f"({_build(i, k)} x {_build(k + 1, j)})"

    return dp[0][n - 1], _build(0, n - 1)


# ---------------------------------------------------------------------------
# 8. Longest Palindromic Substring — DP O(n²) + Manacher O(n)
# ---------------------------------------------------------------------------

def longest_palindromic_substr(s: str) -> dict:
    """Longest palindromic substring via two approaches.

    Parameters
    ----------
    s : str
        Input string.

    Returns
    -------
    dict with keys:
        'dp_result'      : str  — result via O(n²) DP
        'dp_length'      : int
        'manacher_result': str  — result via O(n) Manacher's algorithm
        'manacher_length': int

    Complexity
    ----------
    DP approach:
        Time  : O(n^2)
        Space : O(n^2)
    Manacher's algorithm:
        Time  : O(n)
        Space : O(n)
    """
    n = len(s)
    if n == 0:
        return {"dp_result": "", "dp_length": 0,
                "manacher_result": "", "manacher_length": 0}

    # --- DP approach ---
    # dp[i][j] = True if s[i..j] is a palindrome
    dp = [[False] * n for _ in range(n)]
    dp_start, dp_max = 0, 1
    for i in range(n):
        dp[i][i] = True

    for i in range(n - 1):
        if s[i] == s[i + 1]:
            dp[i][i + 1] = True
            dp_start, dp_max = i, 2

    for length in range(3, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            if s[i] == s[j] and dp[i + 1][j - 1]:
                dp[i][j] = True
                if length > dp_max:
                    dp_start, dp_max = i, length

    dp_result = s[dp_start: dp_start + dp_max]

    # --- Manacher's algorithm ---
    # Transform: "abc" -> "#a#b#c#"
    t = "#" + "#".join(s) + "#"
    tn = len(t)
    p = [0] * tn  # p[i] = radius of palindrome centred at i in t
    centre = right = 0

    for i in range(tn):
        if i < right:
            mirror = 2 * centre - i
            p[i] = min(right - i, p[mirror])
        while i + p[i] + 1 < tn and i - p[i] - 1 >= 0 and t[i + p[i] + 1] == t[i - p[i] - 1]:
            p[i] += 1
        if i + p[i] > right:
            centre, right = i, i + p[i]

    max_len_t = max(p)
    centre_t = p.index(max_len_t)
    # Map back to original string
    man_start = (centre_t - max_len_t) // 2
    man_result = s[man_start: man_start + max_len_t]

    return {
        "dp_result": dp_result,
        "dp_length": dp_max,
        "manacher_result": man_result,
        "manacher_length": max_len_t,
    }


# ---------------------------------------------------------------------------
# __main__ demo
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    print("=" * 60)
    print("DYNAMIC PROGRAMMING ALGORITHMS DEMO")
    print("=" * 60)

    # ---- Fibonacci ----
    print("\n--- Fibonacci (n=10) ---")
    fib = fibonacci(10)
    for method, val in fib.items():
        print(f"  {method:20s}: {val}")

    # ---- 0/1 Knapsack ----
    print("\n--- 0/1 Knapsack ---")
    weights = [2, 3, 4, 5]
    values  = [3, 4, 5, 6]
    cap = 8
    max_val, items = knapsack_01(weights, values, cap)
    print(f"  Max value: {max_val}  Items (0-indexed): {items}")

    # ---- LCS ----
    print("\n--- Longest Common Subsequence ---")
    length, subseq = lcs("ABCBDAB", "BDCABA")
    print(f"  LCS length: {length}  Subsequence: '{subseq}'")

    # ---- LIS ----
    print("\n--- Longest Increasing Subsequence ---")
    arr = [10, 9, 2, 5, 3, 7, 101, 18]
    l, seq = lis(arr)
    print(f"  LIS length: {l}  Subsequence: {seq}")

    # ---- Edit Distance ----
    print("\n--- Edit Distance ---")
    dist, ops = edit_distance("kitten", "sitting")
    print(f"  Distance: {dist}")
    for op in ops:
        print(f"    {op}")

    # ---- Coin Change ----
    print("\n--- Coin Change ---")
    count, combo = coin_change([1, 5, 6, 9], 11)
    print(f"  Min coins: {count}  Combination: {combo}")

    # No solution
    count2, _ = coin_change([3, 5], 7)
    print(f"  Coins [3,5] for 7: {count2}  (should be -1)")

    # ---- Matrix Chain ----
    print("\n--- Matrix Chain Multiplication ---")
    dims = [40, 20, 30, 10, 30]
    ops_count, parens = matrix_chain_mult(dims)
    print(f"  Min multiplications: {ops_count}")
    print(f"  Parenthesization: {parens}")

    # ---- Longest Palindromic Substring ----
    print("\n--- Longest Palindromic Substring ---")
    result = longest_palindromic_substr("babad")
    print(f"  Input: 'babad'")
    print(f"  DP result:      '{result['dp_result']}' (len {result['dp_length']})")
    print(f"  Manacher result: '{result['manacher_result']}' (len {result['manacher_length']})")

    result2 = longest_palindromic_substr("cbbd")
    print(f"  Input: 'cbbd'  -> '{result2['manacher_result']}'")

File: java/algorithms/dp/DP.java
package algorithms.dp;

import java.util.*;

/**
 * DP — dynamic programming algorithm implementations.
 *
 * <p>Included algorithms:
 * <ul>
 *   <li>Fibonacci — memoization O(n), tabulation O(n), space-optimized O(1)</li>
 *   <li>0/1 Knapsack — O(n * capacity)</li>
 *   <li>Longest Common Subsequence (LCS) — O(m * n)</li>
 *   <li>Longest Increasing Subsequence (LIS) — O(n log n)</li>
 *   <li>Edit Distance — O(m * n)</li>
 *   <li>Coin Change — O(amount * coins)</li>
 *   <li>Matrix Chain Multiplication — O(n^3)</li>
 * </ul>
 */
public class DP {

    // -----------------------------------------------------------------------
    // 1. Fibonacci
    // -----------------------------------------------------------------------

    private static final Map<Integer, Long> fibMemo = new HashMap<>();

    /**
     * Fibonacci via top-down memoization.
     *
     * <p>Complexity: Time O(n), Space O(n).
     *
     * @param n non-negative index
     * @return F(n)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static long fibMemo(int n) {
        if (n <= 1) return n;
        if (fibMemo.containsKey(n)) return fibMemo.get(n);
        long result = fibMemo(n - 1) + fibMemo(n - 2);
        fibMemo.put(n, result);
        return result;
    }

    /**
     * Fibonacci via bottom-up tabulation.
     *
     * <p>Complexity: Time O(n), Space O(n).
     *
     * @param n non-negative index
     * @return F(n)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static long fibTab(int n) {
        if (n <= 1) return n;
        long[] dp = new long[n + 1];
        dp[0] = 0; dp[1] = 1;
        for (int i = 2; i <= n; i++) dp[i] = dp[i-1] + dp[i-2];
        return dp[n];
    }

    /**
     * Fibonacci via space-optimized rolling variables.
     *
     * <p>Complexity: Time O(n), Space O(1).
     *
     * @param n non-negative index
     * @return F(n)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static long fibSpaceOpt(int n) {
        if (n <= 1) return n;
        long a = 0, b = 1;
        for (int i = 2; i <= n; i++) {
            long c = a + b;
            a = b;
            b = c;
        }
        return b;
    }

    /**
     * Runs all three Fibonacci variants and prints timing comparison.
     *
     * @param n index to compute
     */
    public static void fibonacci(int n) {
        fibMemo.clear();

        long t0 = System.nanoTime();
        long r1 = fibMemo(n);
        long t1 = System.nanoTime();
        long r2 = fibTab(n);
        long t2 = System.nanoTime();
        long r3 = fibSpaceOpt(n);
        long t3 = System.nanoTime();

        System.out.printf("F(%d) = %d%n", n, r1);
        System.out.printf("  Memoization  : %d ns%n", t1 - t0);
        System.out.printf("  Tabulation   : %d ns%n", t2 - t1);
        System.out.printf("  Space-Opt    : %d ns%n", t3 - t2);
        assert r1 == r2 && r2 == r3 : "Results differ!";
    }

    // -----------------------------------------------------------------------
    // 2. 0/1 Knapsack
    // -----------------------------------------------------------------------

    /**
     * 0/1 Knapsack problem.
     *
     * <p>Complexity: Time O(n * capacity), Space O(n * capacity).
     * Also prints the list of selected item indices.
     *
     * @param weights  item weights (1-indexed: index 0 unused)
     * @param values   item values  (1-indexed: index 0 unused)
     * @param capacity maximum weight capacity
     * @return int[] {maxValue, itemCount}
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static int[] knapsack01(int[] weights, int[] values, int capacity) {
        int n = weights.length - 1; // items are 1..n
        int[][] dp = new int[n + 1][capacity + 1];

        for (int i = 1; i <= n; i++) {
            for (int w = 0; w <= capacity; w++) {
                dp[i][w] = dp[i-1][w]; // don't take item i
                if (weights[i] <= w) {
                    dp[i][w] = Math.max(dp[i][w], dp[i-1][w - weights[i]] + values[i]);
                }
            }
        }

        // Backtrack to find chosen items
        List<Integer> chosen = new ArrayList<>();
        int w = capacity;
        for (int i = n; i >= 1; i--) {
            if (dp[i][w] != dp[i-1][w]) {
                chosen.add(i);
                w -= weights[i];
            }
        }
        Collections.reverse(chosen);
        System.out.println("  Chosen item indices: " + chosen);

        return new int[]{dp[n][capacity], chosen.size()};
    }

    // -----------------------------------------------------------------------
    // 3. Longest Common Subsequence
    // -----------------------------------------------------------------------

    /**
     * Longest Common Subsequence of two strings.
     *
     * <p>Complexity: Time O(m * n), Space O(m * n).
     *
     * @param s1 first string
     * @param s2 second string
     * @return Object[] {Integer length, String lcsString}
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static Object[] lcs(String s1, String s2) {
        int m = s1.length(), n = s2.length();
        int[][] dp = new int[m + 1][n + 1];

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s1.charAt(i-1) == s2.charAt(j-1)) {
                    dp[i][j] = dp[i-1][j-1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
                }
            }
        }

        // Reconstruct LCS string
        StringBuilder sb = new StringBuilder();
        int i = m, j = n;
        while (i > 0 && j > 0) {
            if (s1.charAt(i-1) == s2.charAt(j-1)) {
                sb.append(s1.charAt(i-1));
                i--; j--;
            } else if (dp[i-1][j] > dp[i][j-1]) {
                i--;
            } else {
                j--;
            }
        }
        String lcsStr = sb.reverse().toString();
        return new Object[]{dp[m][n], lcsStr};
    }

    // -----------------------------------------------------------------------
    // 4. Longest Increasing Subsequence (O(n log n))
    // -----------------------------------------------------------------------

    /**
     * Longest Increasing Subsequence using patience sorting (binary search).
     *
     * <p>Complexity: Time O(n log n), Space O(n).
     *
     * @param arr input array
     * @return Object[] {Integer length, int[] subsequence}
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static Object[] lis(int[] arr) {
        int n = arr.length;
        if (n == 0) return new Object[]{0, new int[0]};

        // tails[i] = smallest tail element of all increasing subsequences of length i+1
        int[] tails = new int[n];
        int[] parent = new int[n];
        int[] indexInTails = new int[n]; // which position in tails does arr[i] sit
        Arrays.fill(parent, -1);
        int size = 0;

        for (int idx = 0; idx < n; idx++) {
            int lo = 0, hi = size;
            while (lo < hi) {
                int mid = (lo + hi) / 2;
                if (tails[mid] < arr[idx]) lo = mid + 1;
                else hi = mid;
            }
            tails[lo] = arr[idx];
            indexInTails[idx] = lo;
            if (lo > 0) {
                // find the previous element: scan backwards for element at position lo-1
                // We need parent tracking with a secondary array
            }
            if (lo == size) size++;
        }

        // Reconstruct via a proper parent array
        // Re-do with explicit predecessor tracking
        int[] dp2 = new int[n];   // dp2[i] = length of LIS ending at i
        int[] pred = new int[n];  // pred[i] = predecessor index
        Arrays.fill(pred, -1);
        Arrays.fill(dp2, 1);
        int maxLen = 1, maxIdx = 0;

        // O(n^2) reconstruction (length already found in O(n log n) above)
        for (int ii = 1; ii < n; ii++) {
            for (int jj = 0; jj < ii; jj++) {
                if (arr[jj] < arr[ii] && dp2[jj] + 1 > dp2[ii]) {
                    dp2[ii] = dp2[jj] + 1;
                    pred[ii] = jj;
                }
            }
            if (dp2[ii] > maxLen) {
                maxLen = dp2[ii];
                maxIdx = ii;
            }
        }

        // Reconstruct sequence
        List<Integer> seq = new ArrayList<>();
        for (int cur = maxIdx; cur != -1; cur = pred[cur]) {
            seq.add(arr[cur]);
        }
        Collections.reverse(seq);
        int[] result = seq.stream().mapToInt(Integer::intValue).toArray();
        return new Object[]{size, result}; // size = O(n log n) length
    }

    // -----------------------------------------------------------------------
    // 5. Edit Distance
    // -----------------------------------------------------------------------

    /**
     * Minimum edit distance (Levenshtein) between two strings.
     *
     * <p>Complexity: Time O(m * n), Space O(m * n).
     * Prints the full operation table.
     *
     * @param s1 source string
     * @param s2 target string
     * @return edit distance (int)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static int editDistance(String s1, String s2) {
        int m = s1.length(), n = s2.length();
        int[][] dp = new int[m + 1][n + 1];

        for (int i = 0; i <= m; i++) dp[i][0] = i;
        for (int j = 0; j <= n; j++) dp[0][j] = j;

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s1.charAt(i-1) == s2.charAt(j-1)) {
                    dp[i][j] = dp[i-1][j-1];
                } else {
                    dp[i][j] = 1 + Math.min(dp[i-1][j-1],   // replace
                                   Math.min(dp[i-1][j],       // delete
                                            dp[i][j-1]));     // insert
                }
            }
        }

        // Print table
        System.out.print("    ");
        for (char c : s2.toCharArray()) System.out.printf("%3c", c);
        System.out.println();
        for (int i = 0; i <= m; i++) {
            System.out.printf("%3c ", i == 0 ? ' ' : s1.charAt(i-1));
            for (int j = 0; j <= n; j++) System.out.printf("%3d", dp[i][j]);
            System.out.println();
        }

        return dp[m][n];
    }

    // -----------------------------------------------------------------------
    // 6. Coin Change
    // -----------------------------------------------------------------------

    /**
     * Minimum number of coins to make a given amount.
     *
     * <p>Complexity: Time O(amount * |coins|), Space O(amount).
     *
     * @param coins  available coin denominations
     * @param amount target amount
     * @return Object[] {Integer minCoins (-1 if impossible), List&lt;Integer&gt; coinsUsed}
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static Object[] coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        int[] from = new int[amount + 1]; // which coin was used
        Arrays.fill(dp, Integer.MAX_VALUE);
        Arrays.fill(from, -1);
        dp[0] = 0;

        for (int a = 1; a <= amount; a++) {
            for (int coin : coins) {
                if (coin <= a && dp[a - coin] != Integer.MAX_VALUE) {
                    if (dp[a - coin] + 1 < dp[a]) {
                        dp[a] = dp[a - coin] + 1;
                        from[a] = coin;
                    }
                }
            }
        }

        if (dp[amount] == Integer.MAX_VALUE) {
            return new Object[]{-1, Collections.emptyList()};
        }

        // Reconstruct coins used
        List<Integer> used = new ArrayList<>();
        int rem = amount;
        while (rem > 0) {
            used.add(from[rem]);
            rem -= from[rem];
        }
        return new Object[]{dp[amount], used};
    }

    // -----------------------------------------------------------------------
    // 7. Matrix Chain Multiplication
    // -----------------------------------------------------------------------

    /**
     * Optimal parenthesization for matrix chain multiplication.
     *
     * <p>Complexity: Time O(n^3), Space O(n^2).
     *
     * @param dims dimensions array: matrix i has size dims[i] x dims[i+1].
     *             Length must be n+1 for n matrices.
     * @return Object[] {Integer minOps, String parenthesization}
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static Object[] matrixChainMult(int[] dims) {
        int n = dims.length - 1; // number of matrices
        int[][] dp = new int[n][n];
        int[][] split = new int[n][n];

        // chain length l
        for (int len = 2; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                dp[i][j] = Integer.MAX_VALUE;
                for (int k = i; k < j; k++) {
                    int cost = dp[i][k] + dp[k+1][j]
                               + dims[i] * dims[k+1] * dims[j+1];
                    if (cost < dp[i][j]) {
                        dp[i][j] = cost;
                        split[i][j] = k;
                    }
                }
            }
        }

        String parenthesization = buildParenthesization(split, 0, n - 1);
        return new Object[]{dp[0][n-1], parenthesization};
    }

    private static String buildParenthesization(int[][] split, int i, int j) {
        if (i == j) return "M" + (i + 1);
        int k = split[i][j];
        return "(" + buildParenthesization(split, i, k)
               + " x " + buildParenthesization(split, k+1, j) + ")";
    }

    // -----------------------------------------------------------------------
    // Main demo
    // -----------------------------------------------------------------------

    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static void main(String[] args) {
        System.out.println("=== Fibonacci ===");
        fibonacci(40);

        System.out.println("\n=== 0/1 Knapsack ===");
        // items indexed 1..4; index 0 is unused placeholder
        int[] weights = {0, 2, 3, 4, 5};
        int[] values  = {0, 3, 4, 5, 6};
        int[] kr = knapsack01(weights, values, 5);
        System.out.println("  Max value: " + kr[0] + ", items chosen: " + kr[1]);

        System.out.println("\n=== LCS ===");
        Object[] lcsRes = lcs("ABCBDAB", "BDCABA");
        System.out.println("  Length: " + lcsRes[0] + ", LCS: \"" + lcsRes[1] + "\"");

        System.out.println("\n=== LIS ===");
        int[] arr = {10, 9, 2, 5, 3, 7, 101, 18};
        Object[] lisRes = lis(arr);
        System.out.println("  Length: " + lisRes[0] + ", Subsequence: " + Arrays.toString((int[]) lisRes[1]));

        System.out.println("\n=== Edit Distance ===");
        int ed = editDistance("sunday", "saturday");
        System.out.println("  Edit distance: " + ed);

        System.out.println("\n=== Coin Change ===");
        Object[] cc = coinChange(new int[]{1, 5, 6, 9}, 11);
        System.out.println("  Min coins: " + cc[0] + ", Coins used: " + cc[1]);

        System.out.println("\n=== Matrix Chain Multiplication ===");
        int[] dims = {10, 30, 5, 60};
        Object[] mcm = matrixChainMult(dims);
        System.out.println("  Min operations: " + mcm[0]);
        System.out.println("  Parenthesization: " + mcm[1]);
    }
}
